### ISSUE TITLE
Incorrect type inference on tuple union

### ISSUE THREAD

**Issue by laggron42** [microsoft/pyright#9324]
**Describe the bug**

Pyright is unable to narrow a tuple with literals to the correct type.

**Code or Screenshots**

In the example below, the return type of a function is a union of tuples, one with any boolean and an int, the other with `False` and an error string.

When checking if `a` is True, the type of `b` should be narrowed to `int`, as the error string can only appear if `a` is False.

```python
type t = tuple[bool, int] | tuple[Literal[False], Literal["Something failed"]]

def get() -> t:
    return (True, 42)

a, b = get()

if a:
    reveal_type(b)  # Type of "b" is "int | Literal['Something failed']"
```

**VS Code extension or command-line**

Command line, Pyright version 1.1.386, Python 3.13.0

**Comment by laggron42** [comment]
Hi Eric, thanks for the explainations!

I understand why the additional issues are intended design, thanks for the details. However, you forgot to mention my initial point, that is, the case when there's a union between `bool` and `Literal[False]`.

```py
type t = tuple[bool, int] | tuple[Literal[False], Literal["Something failed"]]

def get() -> t:
    return (True, 42)

x = get()

if x[0] == True:
    reveal_type(x[1])  # Type of "x[1]" is "int | Literal['Something failed']"
```
Here, when `x[0]` equals to `True`, it cannot be false, so it should rule out the branch where `x[0]` is `Literal[False]` and the type of `x[1]` should be `int`.

Sorry if I missed a point, but that one doesn't seem to be the intended design, is it?

**Comment by erictraut** [comment]
The type narrowing logic for the `x[I] == V` type guard pattern is intended to distinguish between tuples that use literal types in their definitions. It doesn't apply to non-literal types like `bool`. If you want type narrowing to work in this case, you would need to change your code to this:

```python
type t = tuple[Literal[True], int] | tuple[Literal[False], Literal["Something failed"]]
```
